test/unit/test_cleanup.py

```diff
@@ -0,0 +1,85 @@
+import os
+import random
+import time
+
+import pytest
+
+from ansible_runner.cleanup import cleanup_dirs, validate_pattern
+from ansible_runner.config.runner import RunnerConfig
+
+
+def test_simple_dir_cleanup_with_exclusions(tmp_path):
+    paths = []
+    for i in range(0, 6, 2):
+        trailing = ''.join(random.choice("abcdefica3829") for i in range(8))
+        path = tmp_path / f'pattern_{i}_{trailing}'
+        path.mkdir()
+        paths.append(path)
+
+    a_file_path = tmp_path / 'pattern_32_donotcleanme'
+    a_file_path.write_text('this is a file and should not be cleaned by the cleanup command')
+
+    keep_dir_path = tmp_path / 'pattern_42_alsokeepme'
+    keep_dir_path.mkdir()
+
+    ct = cleanup_dirs(pattern=str(tmp_path / 'pattern_*_*'), exclude_strings=[42], grace_period=0)
+    assert ct == 3  # cleaned up 3 private_data_dirs
+
+    for path in paths:
+        assert not path.exists()
+
+    assert a_file_path.exists()
+    assert keep_dir_path.exists()
+
+    assert cleanup_dirs(pattern=str(tmp_path / 'pattern_*_*'), exclude_strings=[42], grace_period=0) == 0  # no more to cleanup
+
+
+def test_cleanup_command_grace_period(tmp_path):
+    old_dir = str(tmp_path / 'modtime_old_xyz')
+    new_dir = str(tmp_path / 'modtime_new_abc')
+    os.mkdir(old_dir)
+    time.sleep(1)
+    os.mkdir(new_dir)
+    ct = cleanup_dirs(pattern=str(tmp_path / 'modtime_*_*'), grace_period=1. / 60.)
+    assert ct == 1
+    assert not os.path.exists(old_dir)
+    assert os.path.exists(new_dir)
```

Comment:
You could keep these as `Path` objects.
```suggestion
    old_dir = tmp_path / 'modtime_old_xyz'
    new_dir = tmp_path / 'modtime_new_abc'
    
    old_dir.mkdir()
    time.sleep(1)
    new_dir.mkdir()
    
    ct = cleanup_dirs(pattern=str(tmp_path / 'modtime_*_*'), grace_period=1. / 60.)
    
    assert ct == 1
    assert not old_dir.exists()
    assert new_dir.exists()
```

Side note: using `time.sleep()` in tests is usually asking for instability. It's better to just set a/c/mtimes explicitly rather than slow down the tests unnecessary.